synapse/storage/databases/main/user_directory.py

```diff
@@ -750,6 +755,29 @@ async def search_user_dir(self, user_id, search_term, limit):
                 )
             """
 
+        # We allow manipulating the ranking algorithm by injecting statements
+        # based on config options.
+        additional_ordering_statements = []
+        ordering_arguments = ()
+
+        # If enabled, this config option will rank local users higher than those on
+        # remote instances.
+        if self._prefer_local_users_in_search:
+            # The statement checks whether a given user's user ID contains a domain name
+            # that matches the local server
+            if isinstance(self.database_engine, PostgresEngine):
+                statement = "* (CASE WHEN user_id LIKE ? THEN 2.0 ELSE 1.0 END)"
+            elif isinstance(self.database_engine, Sqlite3Engine):
+                # Note that we need to include a comma at the end for valid SQL
+                statement = "user_id LIKE ? DESC,"
+            else:
+                # This should be unreachable.
+                raise Exception("Unrecognized database engine")
+            additional_ordering_statements.append(statement)
+
+            # Append the local server name as an argument to the final query
+            ordering_arguments += ("%:" + self._server_name,)
```

Comment:
I think it would probably make more sense to split this in two and put into the db specific branches below, otherwise matching up the query structures is quite confusing

Reply:
Done in e87885d81. I agree that it does look a bit cleaner, and will be nicer if we ever (heaven forbid) add another DB engine implementation.